### PR TITLE
Bluetooth: Mesh: Light LC Server models fixes

### PR DESCRIPTION
**Pull request by PavelVPV**
Summary of the fixes (each fix in a separate commit):
- Fixed binding between Light LC Light OnOff and Generic OnOff states
- Internal sum of PI Feedback Regulator is reset when Light LC State Machine is in Off state
- Restoring Light Lightness model states after power up is moved to a separate function
- Added Light LC Light OnOff state publication on power up and after Light LC State Machine transitions to Off state

### CODE REVIEW

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/light_ctrl_srv.c:250]: Seems unnecessary to move this out of the else, now it'll just be overwritten afterwards, and same with the target_on_off. Can this be moved back inside the else?

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/light_ctrl_srv.c:1324]: This breaks the fallback to the configured transition time in `turn_on` and `turn_off`. IIRC the transition will never be null here.

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/light_ctrl_srv.c:1555]: Nice improvement!

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Comment by trond-snekvik** [comment]
Please prefix your commit (and the PR title) with Bluetooth: Mesh:.

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (PavelVPV) [reply subsys/bluetooth/mesh/light_ctrl_srv.c:1324]: Hmm, ok, I initially thought that rules for the transition time for Generic OnOff state should be applied here, but after reading this part of the spec (which got more clarification), seems like here we should either use the transition time provided in the message or the default one for Light LC:
```
If the transition is started as a result of receiving the Light LC Light OnOff Set or Light LC Light OnOff Set Unacknowledged message (see Section 6.5.1.6.2), or as a result of receiving the Generic OnOff Set or Generic OnOff Set Unacknowledged message, changing the bound Generic OnOff state (see Section 6.2.3.3.1), and the Transition Time field is present in this message, it shall be used as the Transition Time.
```

However, Generic OnOff Server will always provide transition time set to either one that was provided in the message or to Generic Default Transition Time state: https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/bluetooth/mesh/gen_onoff_srv.c#L81-L85 So if the default transition time is set in Generic Default Transition Time Server (greater than 0), we will still use this time, but not the default one for Light LC.

File: subsys/bluetooth/mesh/lightness_srv.c
Comment (alxelax) [inline subsys/bluetooth/mesh/lightness_srv.c:835]: Is this comment still relevant?

File: subsys/bluetooth/mesh/lightness_srv.c
Comment (alxelax) [inline subsys/bluetooth/mesh/lightness_srv.c:859]: Is it really required to return value? Since it is an internal API probably it is worth simplifying the prototype.
As a recommendation, I do not insist.

**Comment by alxelax** [comment]
I guess it requires labels `CI-ble-test` and `bug`.  How about squashing commits? @trond-snekvik ?

**Comment by trond-snekvik** [comment]
The commits all seem like reasonable units of change, no need to squash imo

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (alxelax) [inline subsys/bluetooth/mesh/light_ctrl_srv.c:489]: https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fsdk-nrf/detail/PR-3827/4/tests

File: subsys/bluetooth/mesh/lightness_srv.c
Comment (PavelVPV) [reply subsys/bluetooth/mesh/lightness_srv.c:859]: The return value is used in `bt_mesh_lightness_srv_start`.

File: subsys/bluetooth/mesh/lightness_srv.c
Comment (joerchan) [inline subsys/bluetooth/mesh/lightness_srv.c:682]: Did this change end up in the wrong commit?

The commit message does not explain this change.

File: subsys/bluetooth/mesh/light_ctrl_srv.c
Comment (joerchan) [inline subsys/bluetooth/mesh/light_ctrl_srv.c:274]: (void) return value?

**Review by alxelax — APPROVED** [review]

File: subsys/bluetooth/mesh/lightness_srv.c
Comment (PavelVPV) [reply subsys/bluetooth/mesh/lightness_srv.c:682]: You are right, this should be in a separate commit. Thanks for catching this!

**Review by trond-snekvik — APPROVED** [review]

**Comment by PavelVPV** [comment]
I've added a commit that allows to send Generic OnOff Set message to LC element without Transition Time. In that case, transition times defined by the Light LC Property states will be used.

**Comment by trond-snekvik** [comment]
Please resolve compliance issues.

I believe this last commit will need a line in the release notes, it's a behavior change in the API. 
This change should also be implemented in all models that present a transition time pointer, and don't extend the DTT server. 
That should be a separate PR, so we can get this in. I don't think the unilateral change needs to make it into the release though, there's no usecase for it in the other models, so it's just a consistency improvement.

**Comment by PavelVPV** [comment]
If I see correctly, we don't need this change for other models. This is a special case for Light LC because its transition time rules deviates from the standart behaviour. For all other models the transition should be instantaneous if DTT is not present on the element and Transition Time is not included into a message. In this case `bt_mesh_dtt_srv_transition_get` will set transition parameters to `0` which is compliant with the spec.

So far there are only 2 models that extend Generic OnOff Server: Generic Power OnOff Server and Light LC Server. Generic Power OnOff Server instantiates DTT, so it shouldn't cause a problem to the models that extend it.

File: include/bluetooth/mesh/gen_onoff.h
Comment (alxelax) [inline include/bluetooth/mesh/gen_onoff.h:40]: Should Scheduler Server set this pointer to NULL if the scheduled transition time is equal to 0?
It will emulate SetUnack for any Generic OnOff server including that extends LC.

File: subsys/bluetooth/mesh/gen_onoff_srv.c
Comment (alxelax) [inline subsys/bluetooth/mesh/gen_onoff_srv.c:87]: Hm. Might still DTT be present as part of another complex model? For example, the Scene Setup server extends the Generic Default Transition Time Server model. The specification does not prohibit keeping Scene and LC on the same element.

I guess it requires a more complex analysis of the binary tree of extensions.

File: subsys/bluetooth/mesh/gen_ponoff_srv.c
Comment (alxelax) [inline subsys/bluetooth/mesh/gen_ponoff_srv.c:260]: Missed 0 in brackets? if no why is it needed?

**Comment by alxelax** [comment]
It is a bit weird GitHub review system. If I approved it before some changes it makes approved automatically if I choose the comments stage after one more review round.  It is approved by me but I'd like to get answers to my questions :)

**Comment by PavelVPV** [comment]
I've changed the last commit to avoid the issue that @alxelax has pointed out.

File: include/bluetooth/mesh/gen_onoff.h
Comment (PavelVPV) [reply include/bluetooth/mesh/gen_onoff.h]: This now is changed. Please see the new version.

File: subsys/bluetooth/mesh/gen_onoff_srv.c
Comment (PavelVPV) [reply subsys/bluetooth/mesh/gen_onoff_srv.c]: Thanks for noticing that! I've proposed another solution. Please check the latest commit.

File: subsys/bluetooth/mesh/gen_ponoff_srv.c
Comment (PavelVPV) [reply subsys/bluetooth/mesh/gen_ponoff_srv.c]: It is assumed that the pointer to the transition parameters in the set struct is never NULL.

**Review by alxelax — APPROVED** [review]
Now it looks Ok for me. Thanks.

File: include/bluetooth/mesh/model_types.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/model_types.h:39]: I prefer the `NULL` version, over this, to be honest. It parses as a proper English sentence in the set-messages (e.g. `if (set->transition) {  gradual_change(); }`), and didn't introduce any confusing parameters in the client model interfaces. The `NULL` check requirement was handled nicely by your `transition_remaining()` function, too, so as long as we use that in the samples, I don't think it's problematic. Is there some usage scenario that can't work with the NULL pointer approach?

Comment (PavelVPV) [reply include/bluetooth/mesh/model_types.h:39]: This fixes the issue that @alxelax has pointed out. Nothing prevents a user to instantiate Generic DTT server on Light LC element (through Scene Server for example). Then, the previous approach won't work, because `bt_mesh_dtt_srv_transition_get` will set transition time.

File: subsys/bluetooth/mesh/gen_onoff_srv.c
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/gen_onoff_srv.c]: This function doesn't depend on extension bindings at all, it will search through this element's models to find the DTT server. If it's present on the element, this function will find it.

File: include/bluetooth/mesh/model_types.h
Comment (trond-snekvik) [reply include/bluetooth/mesh/model_types.h:39]: I unresolved that conversation and added a comment to it. I don't think the issue Alex mentions is valid, but I can't see what you changed there anymore.

Comment (alxelax) [reply include/bluetooth/mesh/model_types.h:39]: Squashing of commits does not keep the history of changes. There are no changes anymore that have been commented on. Nevertheless, I still guess the issue that I emphasized was valid. It is a bad idea to make the assumption in some internal model algorithm based on the presence or absence third model. The combination of models might be arbitrary because of extensions even to not consider custom usage of generic models.

Comment (trond-snekvik) [reply include/bluetooth/mesh/model_types.h:39]: I had a call with Pavel, and he explained the issue. I misunderstood the problem, but I get it now, and you're right. Pavel is doing some investigation to see if we can solve this without affecting the entire model API

File: subsys/bluetooth/mesh/gen_onoff_internal.h
Comment (alxelax) [inline subsys/bluetooth/mesh/gen_onoff_internal.h:2]: Why 2019?

**Review by alxelax — APPROVED** [review]

**Review by trond-snekvik — APPROVED** [review]

**Review by joerchan — APPROVED** [review]

**Pull request merged**
